### PR TITLE
ncm-ncd: run from temporary directory

### PR DESCRIPTION
**Pull request by stdweird**
Replaces #121

### CODE REVIEW

**Comment by stdweird** [comment]
@ttyS4 i haven't tested this, and i still need to add unittests

**Comment by stdweird** [comment]
tested without chroot
@ned21 @ttyS4 does MS use chroot?

**Comment by ttyS4** [comment]
No, we dont use the chroot feature (just confirmed with someone here).

**Comment by ttyS4** [comment]
As far as I see you use the internal directory function here.
One thing which made me not use it is a confusion: if ncm-ncd is called in noaction mode would it still create the directory. I think it should create the directory regardless.

**Comment by stdweird** [comment]
@ttyS4 it will still do it but i'll check (and even if it didn't you can pass `keeps_state => 1` option to ignore noaction for that specific call

**Comment by stdweird** [comment]
as expected it is created (the noaction is only set when handling components, th etempdir is created before that
```
[DEBUG] directory: (tempdir) noaction is false
[VERB] Created temporary directory with tempdir: /tmp/ncm-ncd-XJ9HrT
[DEBUG] Created temp directory /tmp/ncm-ncd-XJ9HrT
```

File: src/main/perl/CLI.pm [hunk 1]
```diff
@@ -501,12 +516,12 @@ sub action
         $self->option("post-hook"), $self->option("post-hook-timeout"),
         );
 
-    if ($self->option("chroot")) {
-        chroot($self->option("chroot")) or die "Unable to chroot to ", $self->option("chroot");
+    if ($chroot) {
+        chroot($chroot) or die "Unable to chroot to $chroot";
```
Comment (ttyS4) [inline src/main/perl/CLI.pm]: Should this be `$self->error` and `$self->finish(-1)` ?

File: src/main/perl/CLI.pm [hunk 2]
```diff
@@ -501,12 +516,12 @@ sub action
...
     }
 
-    if(! chdir($NCD_CHDIR)) {
-        $self->warn("Failed to change to directory $NCD_CHDIR");
+    if (!chdir($run_from)) {
+        $self->warn("Failed to change to directory $run_from: $!");
```
Comment (ttyS4) [inline src/main/perl/CLI.pm]: Should die (actually `$self->finish(-1)`) here?

**Comment by ttyS4** [comment]
Sorry for making you wait long on this.

Other than making it fail if we can't enter tmpdir this looks good to me.
My quick testing shows no errors.

Restoring chdir after each module exec seems unnecessary for me as we chdir before each module execution still probably does not hurt.

**Comment by stdweird** [comment]
@ttyS4 addressed your remarks, unittests pass (but not yet tested)

**Comment by ttyS4** [comment]
Another quick check (without chroot) shows no errors.

**Comment by jrha** [comment]
@stdweird which PR does this replace? Right now it references itself...

**Comment by ttyS4** [comment]
#121

On January 15, 2018 5:18:35 PM GMT+01:00, James Adams <notifications@github.com> wrote:
>@stdweird which PR does this replace? Right now it references itself...

-- 
Csillag Tamás 
mobilról

**Comment by jrha** [comment]
Thanks!

**Comment by stdweird** [comment]
@jrha @ttyS4 sorry, yes this replaces #121

**Comment by jrha** [comment]
@ttyS4 my concern about including this in the 17.12 release is down to the combination of how late the release is and how little I understand the potential impacts of this change.

File: src/main/perl/CLI.pm [hunk 3]
```diff
@@ -501,12 +516,20 @@ sub action
         $self->option("post-hook"), $self->option("post-hook-timeout"),
         );
 
-    if ($self->option("chroot")) {
-        chroot($self->option("chroot")) or die "Unable to chroot to ", $self->option("chroot");
+    if ($chroot) {
+        if (chroot($chroot)) {
```
Comment (jrha) [inline src/main/perl/CLI.pm]: This spews out:
> `[VERB] Perl warning: Insecure dependency in chroot while running with -t switch at /usr/lib/perl/NCD/CLI.pm line 546.`

File: src/main/perl/CLI.pm [hunk 4]
```diff
@@ -501,12 +516,20 @@ sub action
...
+            $self->verbose("Chrooted to $chroot");
+        } else {
+            $self->error("Unable to chroot to $chroot");
+            $self->finish(-1);
+        }
     }
 
-    if(! chdir($NCD_CHDIR)) {
-        $self->warn("Failed to change to directory $NCD_CHDIR");
+    if (chdir($run_from)) {
```
Comment (jrha) [inline src/main/perl/CLI.pm]: This spews out:
> `[VERB] Perl warning: Insecure dependency in chdir while running with -t switch at /usr/lib/perl/NCD/CLI.pm line 554.`

**Comment by stdweird** [comment]
@jrha and it didn't spew anything in current master version?

**Comment by jrha** [comment]
Probably did, not complaining.

**Comment by jrha** [comment]
I've tested this enough to make myself happier, so in it goes...

**Pull request merged**